### PR TITLE
tcp,pipe: fix bind or listen after close handle

### PR DESCRIPTION
**Pull request by theanarkh**
return `UV_EINVAL` in `bind` and `listen` when `handle` is `UV_HANDLE_CLOSING` or `UV_HANDLE_CLOSED`.

fix https://github.com/libuv/libuv/issues/3503

### CODE REVIEW

File: test/test-pipe-bind-error.c [hunk 1]
```diff
@@ -137,3 +137,23 @@ TEST_IMPL(pipe_listen_without_bind) {
...
+  ASSERT(r == UV_EINVAL);
+
+  r = uv_listen((uv_stream_t*)&server, SOMAXCONN, NULL);
+  ASSERT(r == UV_EINVAL);
+
+  uv_run(uv_default_loop(), UV_RUN_DEFAULT);
+
+  MAKE_VALGRIND_HAPPY();
+  return 0;
+}
```
Comment (cjihrig) [inline test/test-pipe-bind-error.c:159]: Please add a newline at the end of the file.

File: test/test-pipe-bind-error.c [hunk 2]
```diff
@@ -137,3 +137,23 @@ TEST_IMPL(pipe_listen_without_bind) {
   MAKE_VALGRIND_HAPPY();
   return 0;
 }
+
+TEST_IMPL(pipe_bind_or_listen_error_after_close) {
+  uv_pipe_t server;
+  int r;
+
+  r = uv_pipe_init(uv_default_loop(), &server, 0);
```
Comment (cjihrig) [inline test/test-pipe-bind-error.c:145]: There are a number of cases where you're only using the return value in an assertion on the following line. Those can be consolidated a bit. Take a look at `ASSERT_EQ()` and similar assertions in the codebase.

File: test/test-tcp-bind-error.c [hunk 3]
```diff
@@ -297,3 +297,28 @@ TEST_IMPL(tcp_bind_writable_flags) {
...
+  r = uv_loop_init(&loop);
+  ASSERT(r == 0);
+  r = uv_tcp_init(&loop, &tcp);
+  ASSERT(r == 0);
+  uv_close((uv_handle_t *)&tcp, NULL);
+  r = uv_tcp_bind(&tcp, (struct sockaddr *)&addr, 0);
+  ASSERT(r == UV_EINVAL);
+  r = uv_listen((uv_stream_t *)&tcp, 5, NULL);
+  ASSERT(r == UV_EINVAL);
+  (void)uv_run(&loop, UV_RUN_DEFAULT);
```
Comment (cjihrig) [inline test/test-tcp-bind-error.c:321]: I don't think you need the `(void)` here, but you can assert on the return value of `uv_run()`.

**Review by cjihrig — COMMENTED** [review]
ASAN is currently failing in the CI run.

File: test/test-pipe-bind-error.c (hunk 1 above)
Comment (theanarkh) [reply test/test-pipe-bind-error.c]: Thanks, done.

File: test/test-tcp-bind-error.c (hunk 3 above)
Comment (theanarkh) [reply test/test-tcp-bind-error.c]: Done.

File: test/test-pipe-bind-error.c (hunk 2 above)
Comment (theanarkh) [reply test/test-pipe-bind-error.c]: Done.

File: test/test-pipe-bind-error.c [hunk 4]
```diff
@@ -137,3 +137,19 @@ TEST_IMPL(pipe_listen_without_bind) {
   MAKE_VALGRIND_HAPPY();
   return 0;
 }
+
+TEST_IMPL(pipe_bind_or_listen_error_after_close) {
+  uv_pipe_t server;
+
+  ASSERT(uv_pipe_init(uv_default_loop(), &server, 0) == 0);
+  uv_close((uv_handle_t *)&server, NULL);
```
Comment (vtjnash) [inline test/test-pipe-bind-error.c:145]: Style nit (for all new tests)
```suggestion
  uv_close((uv_handle_t*) &server, NULL);
```

Secondly, could you write the new tests with the new `ASSERT_EQ()` macro family?

Comment (theanarkh) [reply test/test-pipe-bind-error.c]: Thanks! done.

**Review by cjihrig — APPROVED** [review]

**Pull request merged**